### PR TITLE
[FLINK-7021] Flink Task Manager hangs on startup if one Zookeeper node is unresolvable

### PR DESCRIPTION
**Pull request by skidder**
Fixes issue FLINK-7021 by adding an `UnhandledErrorListener` implementation to the Task Manager that will shutdown the Task Manager if an unretryable exception is raised while retrieving the Zookeeper leader.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java [hunk 1]
```diff
@@ -196,4 +198,12 @@ protected void handleStateChange(ConnectionState newState) {
 				break;
 		}
 	}
+
+	@Override
+	public void unhandledError(String s, Throwable throwable) {
+		LOG.warn("Unhandled error encountered while retrieving Zookeeper leader.");
+		final RuntimeException e = new RuntimeException(s, throwable);
+		this.leaderListener.handleError(e);
+		throw e;
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/ZooKeeperLeaderRetrievalService.java]: I think it should be enough to call `this.leaderListener.handleError`.

**Comment by StephanEwen** [comment]
@skidder Do you want to follow up on this one? Otherwise, another contributor might take this over.
Betting this fix into 1.4 would be great...

**Comment by tillrohrmann** [comment]
I'll address my comment myself and then merge the PR. Thanks a lot for your contribution @skidder.

**Comment by skidder** [comment]
Sorry for not responding sooner! Thank you @StephanEwen and @tillrohrmann for applying this fix!

**Pull request closed without merging**